src/main/psxsdk/libspu/spu.c

```diff
@@ -111,7 +111,89 @@ s32 _spu_init(s32 arg0) {
     return 0;
 }
 
-INCLUDE_ASM("main/nonmatchings/psxsdk/libspu/spu", _spu_writeByIO);
+extern s32 aWaitDmafClearW;
+extern s32 aWaitWrdyHL;
+
+s32 _spu_writeByIO(s32 addr, s32 size) {
+    volatile s32 sp0;
+    volatile s32 sp4;
+    s32 spustat_prev;
+    s32 spustat;
+    s32 wait_count;
+    s32 num_to_trans;
+    u32 total_size;
+    u16* cur_pos;
+    s32 spustat_cur;
+    s32 less_than_64;
+    s32 num_trans;
+    u16 data;
+    u8 pad[8];
+    u16 spucnt;
+
+    total_size = size;
+    cur_pos = addr;
+    spustat = _spu_RXX->rxx.spustat & 0x7FF;
+    _spu_RXX->rxx.trans_addr = _spu_tsa;
+    WASTE_TIME();
+    less_than_64 = total_size < 0x41;
+    if (total_size != 0) {
+        do {
```

Comment:
You can simplify the `if`/`do`/`while` combo as a single `while`: https://decomp.me/scratch/n6mlC